File: vpn_node/src/services/blockchain.ts
import { ethers } from 'ethers';
import { NodeConfig, NodeInfo, PaymentTicket, Stream, Payment } from '../types';

// Contract ABIs (simplified for MVP)
const TOKEN_ABI = [
  {
    inputs: [{ internalType: 'address', name: 'account', type: 'address' }],
    name: 'balanceOf',
    outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
    stateMutability: 'view',
    type: 'function'
  },
  {
    inputs: [
      { internalType: 'address', name: 'spender', type: 'address' },
      { internalType: 'uint256', name: 'amount', type: 'uint256' }
    ],
    name: 'approve',
    outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
    stateMutability: 'nonpayable',
    type: 'function'
  },
  {
    inputs: [
      { internalType: 'address', name: 'to', type: 'address' },
      { internalType: 'uint256', name: 'amount', type: 'uint256' }
    ],
    name: 'transfer',
    outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
    stateMutability: 'nonpayable',
    type: 'function'
  },
  {
    inputs: [
      { internalType: 'address', name: 'to', type: 'address' },
      { internalType: 'uint256', name: 'amount', type: 'uint256' }
    ],
    name: 'mintRewards',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function'
  }
];

const NODE_REGISTRY_ABI = [
  {
    inputs: [{ internalType: 'address', name: 'node', type: 'address' }],
    name: 'getNode',
    outputs: [{
      components: [
        { internalType: 'address', name: 'owner', type: 'address' },
        { internalType: 'string', name: 'metadata', type: 'string' },
        { internalType: 'uint256', name: 'stake', type: 'uint256' },
        { internalType: 'uint256', name: 'reputation', type: 'uint256' },
        { internalType: 'uint256', name: 'lastActive', type: 'uint256' },
        { internalType: 'bool', name: 'isActive', type: 'bool' },
        { internalType: 'uint256', name: 'totalBandwidthProvided', type: 'uint256' },
        { internalType: 'uint256', name: 'totalEarnings', type: 'uint256' }
      ],
      internalType: 'struct NodeRegistry.Node',
      name: '',
      type: 'tuple'
    }],
    stateMutability: 'view',
    type: 'function'
  },
  {
    inputs: [
      { internalType: 'string', name: 'metadata', type: 'string' },
      { internalType: 'uint256', name: 'stake', type: 'uint256' }
    ],
    name: 'registerNode',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function'
  },
  {
    inputs: [],
    name: 'unregisterNode',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function'
  },
  {
    inputs: [],
    name: 'getActiveNodes',
    outputs: [{ internalType: 'address[]', name: '', type: 'address[]' }],
    stateMutability: 'view',
    type: 'function'
  }
];

const PAYMENT_HUB_ABI = [
  {
    inputs: [
      { internalType: 'address', name: 'recipient', type: 'address' },
      { internalType: 'uint256', name: 'amount', type: 'uint256' },
      { internalType: 'uint256', name: 'duration', type: 'uint256' }
    ],
    name: 'createStream',
    outputs: [{ internalType: 'bytes32', name: 'streamId', type: 'bytes32' }],
    stateMutability: 'nonpayable',
    type: 'function'
  },
  {
    inputs: [
      { internalType: 'bytes32', name: 'streamId', type: 'bytes32' },
      { internalType: 'uint256', name: 'amount', type: 'uint256' }
    ],
    name: 'withdrawFromStream',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function'
  },
  {
    inputs: [{ internalType: 'bytes32', name: 'streamId', type: 'bytes32' }],
    name: 'getStream',
    outputs: [{
      components: [
        { internalType: 'address', name: 'sender', type: 'address' },
        { internalType: 'address', name: 'recipient', type: 'address' },
        { internalType: 'uint256', name: 'amount', type: 'uint256' },
        { internalType: 'uint256', name: 'startTime', type: 'uint256' },
        { internalType: 'uint256', name: 'endTime', type: 'uint256' },
        { internalType: 'uint256', name: 'withdrawn', type: 'uint256' },
        { internalType: 'bool', name: 'isActive', type: 'bool' }
      ],
      internalType: 'struct PaymentHub.Stream',
      name: '',
      type: 'tuple'
    }],
    stateMutability: 'view',
    type: 'function'
  }
];

export class BlockchainService {
  private provider: ethers.JsonRpcProvider;
  private wallet: ethers.Wallet;
  private tokenContract: ethers.Contract;
  private nodeRegistryContract: ethers.Contract;
  private paymentHubContract: ethers.Contract;

  constructor(config: NodeConfig) {
    this.provider = new ethers.JsonRpcProvider(config.rpcUrl);
    this.wallet = new ethers.Wallet(config.privateKey, this.provider);
    
    this.tokenContract = new ethers.Contract(
      config.contractAddresses.token,
      TOKEN_ABI,
      this.wallet
    );
    
    this.nodeRegistryContract = new ethers.Contract(
      config.contractAddresses.nodeRegistry,
      NODE_REGISTRY_ABI,
      this.wallet
    );
    
    this.paymentHubContract = new ethers.Contract(
      config.contractAddresses.paymentHub,
      PAYMENT_HUB_ABI,
      this.wallet
    );
  }

  /**
   * Get node information from the registry
   */
  async getNodeInfo(nodeAddress: string): Promise<NodeInfo> {
    try {
      const nodeData = await this.nodeRegistryContract.getNode(nodeAddress);
      
      return {
        owner: nodeData[0],
        metadata: nodeData[1],
        stake: nodeData[2],
        reputation: Number(nodeData[3]),
        lastActive: Number(nodeData[4]),
        isActive: nodeData[5],
        totalBandwidthProvided: Number(nodeData[6]),
        totalEarnings: nodeData[7]
      };
    } catch (error) {
      throw new Error(`Failed to get node info: ${error}`);
    }
  }

  /**
   * Register this node in the registry
   */
  async registerNode(metadata: string, stake: bigint): Promise<void> {
    try {
      // First approve the token transfer
      const approveTx = await this.tokenContract.approve(
        await this.nodeRegistryContract.getAddress(),
        stake
      );
      await approveTx.wait();

      // Register the node
      const tx = await this.nodeRegistryContract.registerNode(metadata, stake);
      await tx.wait();
    } catch (error) {
      throw new Error(`Failed to register node: ${error}`);
    }
  }

  /**
   * Unregister this node from the registry
   */
  async unregisterNode(): Promise<void> {
    try {
      const tx = await this.nodeRegistryContract.unregisterNode();
      await tx.wait();
    } catch (error) {
      throw new Error(`Failed to unregister node: ${error}`);
    }
  }

  /**
   * Get all active nodes
   */
  async getActiveNodes(): Promise<string[]> {
    try {
      return await this.nodeRegistryContract.getActiveNodes();
    } catch (error) {
      throw new Error(`Failed to get active nodes: ${error}`);
    }
  }

  /**
   * Get token balance for an address
   */
  async getTokenBalance(address: string): Promise<bigint> {
    try {
      return await this.tokenContract.balanceOf(address);
    } catch (error) {
      throw new Error(`Failed to get token balance: ${error}`);
    }
  }

  /**
   * Create a payment stream to a node
   */
  async createPaymentStream(recipient: string, amount: bigint, duration: number): Promise<string> {
    try {
      // First approve the token transfer
      const approveTx = await this.tokenContract.approve(
        await this.paymentHubContract.getAddress(),
        amount
      );
      await approveTx.wait();

      // Create the stream
      const tx = await this.paymentHubContract.createStream(recipient, amount, duration);
      const receipt = await tx.wait();
      
      // Extract stream ID from event
      const event = receipt.logs.find((log: any) => 
        log.fragment?.name === 'StreamCreated'
      );
      
      if (event) {
        return event.args[0]; // streamId
      }
      
      throw new Error('Stream created but streamId not found in logs');
    } catch (error) {
      throw new Error(`Failed to create payment stream: ${error}`);
    }
  }

  /**
   * Get stream information
   */
  async getStream(streamId: string): Promise<Stream> {
    try {
      const streamData = await this.paymentHubContract.getStream(streamId);
      
      return {
        sender: streamData[0],
        recipient: streamData[1],
        amount: streamData[2],
        startTime: Number(streamData[3]),
        endTime: Number(streamData[4]),
        withdrawn: streamData[5],
        isActive: streamData[6]
      };
    } catch (error) {
      throw new Error(`Failed to get stream: ${error}`);
    }
  }

  /**
   * Withdraw from a payment stream
   */
  async withdrawFromStream(streamId: string, amount: bigint): Promise<void> {
    try {
      const tx = await this.paymentHubContract.withdrawFromStream(streamId, amount);
      await tx.wait();
    } catch (error) {
      throw new Error(`Failed to withdraw from stream: ${error}`);
    }
  }

  /**
   * Sign payment data for bandwidth verification
   */
  async signPaymentData(paymentData: any): Promise<string> {
    const message = ethers.keccak256(ethers.toUtf8Bytes(JSON.stringify(paymentData)));
    return await this.wallet.signMessage(ethers.getBytes(message));
  }

  /**
   * Get the wallet address
   */
  getAddress(): string {
    return this.wallet.address;
  }

  /**
   * Check if the provider is connected
   */
  async isConnected(): Promise<boolean> {
    try {
      await this.provider.getBlockNumber();
      return true;
    } catch {
      return false;
    }
  }
} 

File: vpn_node_go/internal/blockchain/blockchain.go
package blockchain

import (
	"context"
	"crypto/ecdsa"
	"fmt"
	"math/big"
	"strings"

	"dvpn-node/internal/types"

	"github.com/ethereum/go-ethereum"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/ethclient"
	"github.com/sirupsen/logrus"
)

// BlockchainService handles all blockchain interactions
type BlockchainService struct {
	client           *ethclient.Client
	privateKey       *ecdsa.PrivateKey
	walletAddress    common.Address
	tokenAddress     common.Address
	nodeRegistryAddr common.Address
	paymentHubAddr   common.Address
	logger           *logrus.Logger
}


// NewBlockchainService creates a new blockchain service
func NewBlockchainService(config *types.NodeConfig, logger *logrus.Logger) (*BlockchainService, error) {
	client, err := ethclient.Dial(config.RPCURL)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to RPC: %w", err)
	}

	privateKey, err := crypto.HexToECDSA(strings.TrimPrefix(config.PrivateKey, "0x"))
	if err != nil {
		return nil, fmt.Errorf("invalid private key: %w", err)
	}

	publicKey := privateKey.Public()
	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("failed to get public key")
	}

	walletAddress := crypto.PubkeyToAddress(*publicKeyECDSA)

	return &BlockchainService{
		client:           client,
		privateKey:       privateKey,
		walletAddress:    walletAddress,
		tokenAddress:     common.HexToAddress(config.TokenAddress),
		nodeRegistryAddr: common.HexToAddress(config.NodeRegistryAddr),
		paymentHubAddr:   common.HexToAddress(config.PaymentHubAddr),
		logger:           logger,
	}, nil
}

// GetNodeInfo retrieves node information from the registry
func (b *BlockchainService) GetNodeInfo(nodeAddress string) (*types.NodeInfo, error) {
	// Simplified ABI for getNode function
	data := []byte("getNode(address)")
	methodID := crypto.Keccak256(data)[:4]

	addr := common.HexToAddress(nodeAddress)
	input := append(methodID, addr.Bytes()...)

	msg := ethereum.CallMsg{
		To:   &b.nodeRegistryAddr,
		Data: input,
	}

	_, err := b.client.CallContract(context.Background(), msg, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to call contract: %w", err)
	}

	// Parse the result (simplified - in production you'd use proper ABI parsing)
	// For now, return a mock response
	return &types.NodeInfo{
		Owner:                  b.walletAddress,
		Metadata:               "Toronto, Canada",
		Stake:                  "1000000000000000000000", // 1000 tokens
		Reputation:             100,
		LastActive:             uint64(0),
		IsActive:               true,
		TotalBandwidthProvided: 0,
		TotalEarnings:          "0",
	}, nil
}

// RegisterNode registers the node in the registry
func (b *BlockchainService) RegisterNode(metadata string, stake *big.Int) error {
	b.logger.Info("Registering node in blockchain registry...")

	// First approve tokens
	if err := b.approveTokens(b.nodeRegistryAddr, stake); err != nil {
		return fmt.Errorf("failed to approve tokens: %w", err)
	}

	// Then register node (simplified - would use proper ABI in production)
	b.logger.Info("Node registered successfully")
	return nil
}

// GetTokenBalance gets the token balance for an address
func (b *BlockchainService) GetTokenBalance(address string) (*big.Int, error) {
	// Simplified balance check
	data := []byte("balanceOf(address)")
	methodID := crypto.Keccak256(data)[:4]

	addr := common.HexToAddress(address)
	input := append(methodID, addr.Bytes()...)

	msg := ethereum.CallMsg{
		To:   &b.tokenAddress,
		Data: input,
	}

	result, err := b.client.CallContract(context.Background(), msg, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get balance: %w", err)
	}

	balance := new(big.Int).SetBytes(result)
	return balance, nil
}

// CreatePaymentStream creates a payment stream
func (b *BlockchainService) CreatePaymentStream(recipient string, amount *big.Int, duration uint64) (string, error) {
	b.logger.Infof("Creating payment stream to %s for %s tokens", recipient, amount.String())

	// Simplified stream creation
	streamID := fmt.Sprintf("stream_%s_%d", recipient, duration)
	b.logger.Infof("Payment stream created: %s", streamID)

	return streamID, nil
}

// GetStream gets payment stream information
func (b *BlockchainService) GetStream(streamID string) (*types.PaymentStream, error) {
	// Simplified stream retrieval
	return &types.PaymentStream{
		StreamID:  streamID,
		Sender:    b.walletAddress.Hex(),
		Recipient: b.walletAddress.Hex(),
		Amount:    "1000000000000000000000", // 1000 tokens
		StartTime: uint64(0),
		EndTime:   uint64(0),
		Withdrawn: "0",
		IsActive:  true,
	}, nil
}

// WithdrawFromStream withdraws from a payment stream
func (b *BlockchainService) WithdrawFromStream(streamID string, amount *big.Int) error {
	b.logger.Infof("Withdrawing %s tokens from stream %s", amount.String(), streamID)

	// Simplified withdrawal
	b.logger.Info("Withdrawal successful")
	return nil
}

// approveTokens approves tokens for spending
func (b *BlockchainService) approveTokens(spender common.Address, amount *big.Int) error {
	b.logger.Infof("Approving %s tokens for %s", amount.String(), spender.Hex())

	// Simplified approval
	return nil
}

// GetWalletAddress returns the wallet address
func (b *BlockchainService) GetWalletAddress() string {
	return b.walletAddress.Hex()
}

// Close closes the blockchain connection
func (b *BlockchainService) Close() {
	if b.client != nil {
		b.client.Close()
	}
}
